be/src/util/dynamic_cache.h

```diff
@@ -237,45 +237,55 @@ class DynamicCache {
 
     // clear all unused *and* expired objects
     void clear_expired() {
-        int64_t now = MonotonicMillis();
-        std::lock_guard<std::mutex> lg(_lock);
-        auto itr = _list.begin();
-        while (itr != _list.end()) {
-            Entry* entry = (*itr);
-            if (entry->_ref == 1 && now >= entry->_expire_ms) {
-                // no usage, can remove
-                _map.erase(entry->key());
-                itr = _list.erase(itr);
-                _object_size--;
-                _size -= entry->_size;
-                if (_mem_tracker) _mem_tracker->release(entry->_size);
-                // TODO(cbl): delete without holding lock
-                delete entry;
-            } else {
-                itr++;
+        std::vector<Entry*> entry_list;
+        {
+            int64_t now = MonotonicMillis();
+            std::lock_guard<std::mutex> lg(_lock);
+            auto itr = _list.begin();
+            while (itr != _list.end()) {
+                Entry* entry = (*itr);
+                if (entry->_ref == 1 && now >= entry->_expire_ms) {
+                    // no usage, can remove
+                    _map.erase(entry->key());
+                    itr = _list.erase(itr);
+                    _object_size--;
+                    _size -= entry->_size;
+                    if (_mem_tracker) _mem_tracker->release(entry->_size);
+                    entry_list.push_back(entry);
+                } else {
+                    itr++;
+                }
             }
         }
```

Comment:
What happens if a thread here retrieves an entry marked as 'expired'? Will it result in a NullPointerException?

Reply:
Entry in `entry_list` can be safe deleted, because it already remove from `_map` and `_list`.

Reply:
No one need it anymore.